crates/compiler/mono/src/code_gen_help/refcount.rs

```diff
@@ -412,6 +378,166 @@ pub fn refcount_reset_proc_body<'a>(
     rc_ptr_stmt
 }
 
+pub fn refcount_resetref_proc_body<'a>(
+    root: &mut CodeGenHelp<'a>,
+    ident_ids: &mut IdentIds,
+    ctx: &mut Context<'a>,
+    layout_interner: &mut STLayoutInterner<'a>,
+    layout: InLayout<'a>,
+    structure: Symbol,
+) -> Stmt<'a> {
+    let rc_ptr = root.create_symbol(ident_ids, "rc_ptr");
+    let rc = root.create_symbol(ident_ids, "rc");
+    let refcount_1 = root.create_symbol(ident_ids, "refcount_1");
+    let is_unique = root.create_symbol(ident_ids, "is_unique");
+    let addr = root.create_symbol(ident_ids, "addr");
+
+    let union_layout = match layout_interner.get(layout) {
+        Layout::Union(u) => u,
+        _ => unimplemented!("Reset is only implemented for UnionLayout"),
+    };
+
+    // Whenever we recurse into a child layout we will want to Decrement
+    ctx.op = HelperOp::Dec;
+    ctx.recursive_union = Some(union_layout);
+    let recursion_ptr = layout_interner.insert(Layout::RecursivePointer(layout));
+
+    // Reset structure is unique. Return a pointer to the allocation.
+    let then_stmt = {
+        let alignment = root.create_symbol(ident_ids, "alignment");
+        let alignment_int = layout_interner
+            .get(layout)
+            .allocation_alignment_bytes(layout_interner, root.target_info);
+        let alignment_expr = Expr::Literal(Literal::Int((alignment_int as i128).to_ne_bytes()));
+        let alloc_addr = root.create_symbol(ident_ids, "alloc_addr");
+        let alloc_addr_expr = Expr::Call(Call {
+            call_type: CallType::LowLevel {
+                op: LowLevel::NumSubWrap,
+                update_mode: UpdateModeId::BACKEND_DUMMY,
+            },
+            arguments: root.arena.alloc([addr, alignment]),
+        });
+
+        Stmt::Let(
+            alignment,
+            alignment_expr,
+            root.layout_isize,
+            root.arena.alloc(
+                //
+                Stmt::Let(
+                    alloc_addr,
+                    alloc_addr_expr,
+                    root.layout_isize,
+                    root.arena.alloc(
+                        //
+                        Stmt::Ret(alloc_addr),
```

Comment:
I missed this. I suspect this will still return an invalid value. Can we make a test that hits this case?

Reply:
decrefs are inserted for lists only in inc_dec (so they never get converted into a resetref). Perhaps there are other places where decrefs for unions are inserted?